pkg/output/output.go

```diff
@@ -14,3 +17,73 @@ type Output interface {
 	Write(ctx context.Context, policyManifest *packaging.PolicyManifest, intermediateJSON []byte, rc *results.ResultCollection, cluster string, timestamp time.Time) error
 	WriteIndex(ctx context.Context, cluster string, timestamp time.Time, clusterSummary *api.ClusterSummary) error
 }
+
+// TODO: Change config to use new struct
+// type OutputsConfig struct {
+// 	Local  []*LocalOutputConfig
+// 	CLI    []*CLIOutputConfig
+// 	GCS    []*gcs.GCSOutputConfig
+// 	AZBlob []*azblob.AZBlobOutputConfig
+// }
+
+type OutputConfig struct {
+	Type   string
+	Format string
+	// For local
+	Path string
+	// For GCS
+	BucketName      string
+	CredentialsPath string
+	// For AZBlob
+	ContainerName string
+	AccountName   string
+	AccountKey    string
+}
+
+func NewOutputs(ctx context.Context, config []*OutputConfig) []Output {
+	outputs := make([]Output, 0)
+	for _, outputConfig := range config {
+		if outputConfig.Type == "local" {
+			output, err := NewLocalOutput(ctx, &LocalOutputConfig{
+				Format: outputConfig.Format,
+				Path:   outputConfig.Path,
+			})
+			if err != nil {
+				log.Fatalf("Could not create local output: %s", err)
+			}
+			outputs = append(outputs, output)
+		} else if outputConfig.Type == "cli" {
+			output, err := NewCLIOutput(ctx, &CLIOutputConfig{
+				Format: outputConfig.Format,
+			})
+			if err != nil {
+				log.Fatalf("Could not create CLI output: %s", err)
+			}
+			outputs = append(outputs, output)
+		} else if outputConfig.Type == "gcs" {
+			output, err := gcs.NewGCSOutput(ctx, &gcs.GCSOutputConfig{
+				Format:          outputConfig.Format,
+				CredentialsPath: outputConfig.CredentialsPath,
+				BucketName:      outputConfig.BucketName,
+			})
+			if err != nil {
+				log.Fatalf("Could not create GCS output: %s", err)
+			}
+			outputs = append(outputs, output)
+		} else if outputConfig.Type == "azblob" {
+			output, err := azblob.NewAZBlobOutput(ctx, &azblob.AZBlobOutputConfig{
+				Format:        outputConfig.Format,
+				ContainerName: outputConfig.ContainerName,
+				AccountName:   outputConfig.AccountName,
+				AccountKey:    outputConfig.AccountKey,
+			})
+			if err != nil {
+				log.Fatalf("Could not create AZBlob output: %s", err)
+			}
+			outputs = append(outputs, output)
+		} else {
+			log.Fatalf("Output type not recognised: %s", outputConfig.Type)
+		}
+	}
+	return outputs
+}
```

Comment:
I think all of this belongs to another package. I'd call it `cmd/checkerconfig`, for instance.

`OutputConfig` is just the format of the specific config file for `preflight check`, so I'd encapsulate everything that is related to that under the same package, close to the cmd itself.

I don't think this specific package should be shared with the new agent configuration, for instance.

Reply:
My intent with keeping the config and config checking for each package inside the package is that it makes it easier to reuse in other commands, so it's not specific to `preflight check`.

That way as we build new components if we want them to be able to write output we can just add the struct to their config struct and it will work the same as `check`.

I don't think this will be used in the `agent` command as it won't be writing output will it?

Reply:
> My intent with keeping the config and config checking for each package inside the package is that it makes it easier to reuse in other commands, so it's not specific to preflight check.

Exactly, and that's fine. +1 to the config structs in the specific data-gatherers.

What I was trying to say is that this method that converts the schema of the config file to a slice of outputs fits more naturally inside the cmd itself, since it is coupled to the format of the config file.

> I don't think this will be used in the agent command as it won't be writing output will it?

The agent won't write outputs.

Neither the agent of the backend should use this `OutputConfig` struct. This is coupled to cmd/check.go and because of that it should not be in a shared package but inside `check.go` or an auxiliar package inside `cmd`.